project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java

```diff
@@ -266,27 +258,104 @@ private void createOpenbisSamples(List<SampleCreation> samplesToRegister) {
-    deleteOpenbisSample(DEFAULT_SPACE_CODE, sampleCode.code());
+  public void deleteAll(ProjectCode projectCode,
+      Collection<life.qbic.projectmanagement.domain.model.sample.Sample> samples) {
+
+    Set<String> sampleCodesToDelete = samples.stream().map(sample -> sample.sampleCode().code())
+        .collect(Collectors.toSet());
+
+    //Fetch samples with potential data - sample search is not working, sorry you have to see this
+    ExperimentFetchOptions fetchOptions = new ExperimentFetchOptions();
+    fetchOptions.withSamplesUsing(fetchSamplesCompletely());
+    for (Experiment experiment : searchExperimentsByProjectCode(projectCode.value(), fetchOptions)) {
+      for (Sample sample : experiment.getSamples()) {
+        String sampleCode = sample.getCode();
+        if (sampleCodesToDelete.contains(sampleCode)) {
+          if (isSampleWithData(List.of(sample))) {
+            throw new SampleNotDeletedException(
+                "Did not delete sample " + sampleCode + ", because data is attached.");
+          }
+        }
+      }
+    }
+    // no data found, we can safely delete all samples
+    sampleCodesToDelete.forEach(code -> deleteOpenbisSample(DEFAULT_SPACE_CODE, code));
+  }
+
+  /**
+   * Recursive method checking child samples for datasets
+   */
+  private boolean isSampleWithData(List<Sample> samples) {
+    boolean hasData = false;
+    for (Sample sample : samples) {
+      hasData |= !sample.getDataSets().isEmpty();
+      hasData |= isSampleWithData(sample.getChildren());
+    }
+    return hasData;
   }
 
   /**
-   * Searches for samples that contain the provided sample code
+   * Updates the reference to one or more {@link Sample}s in the data repository to connect project
+   * data. Samples with metadata must be provided. Since no batch information is stored, changes in
+   * the batch are not reflected.
    *
-   * @param sampleCode the {@link SampleCode} to search for in the data repository
-   * @return true, if a sample with that code already exists in the system, false if not
+   * @param samples the batch of {@link Sample}s to be updated in the data repo
    * @since 1.0.0
    */
   @Override
-  public boolean sampleExists(SampleCode sampleCode) {
-    return !searchSamplesByCode(sampleCode.code()).isEmpty();
+  public void updateAll(
+      Collection<life.qbic.projectmanagement.domain.model.sample.Sample> samples)
+      throws SampleNotUpdatedException {
+    try {
+      updateOpenbisSamples(convertSamplesToSampleUpdates(samples));
+    } catch (Exception e) {
+      throw new SampleNotUpdatedException(
+          "Samples could not be updated due to " + e.getCause() + " with " + e.getMessage());
+    }
+  }
+
+  private List<SampleUpdate> convertSamplesToSampleUpdates(
+      Collection<life.qbic.projectmanagement.domain.model.sample.Sample> updatedSamples) {
+    List<SampleUpdate> updatedSamplesList = new ArrayList<>();
+    try {
+      updatedSamples.forEach(sample -> {
```

Comment:
try to avoid multi line, compley functions as lambda. 
Possible solutions:
* stream over the list twice, each operation doing a separate concern
* extract a method
* use map more and filter and other stream operations instead of a uber-Foreach